amazon-kinesis-client/src/main/java/software/amazon/kinesis/coordinator/PeriodicShardSyncManager.java

```diff
@@ -178,12 +178,20 @@ private void runShardSync() {
             final long runStartMillis = System.currentTimeMillis();
 
             try {
+                // Create a shallow copy of the map to avoid data race with Scheduler.
+                final Map<StreamIdentifier, StreamConfig> streamConfigMap = currentStreamConfigMap.entrySet().stream()
+                        .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
                 // Construct the stream to leases map to be used in the lease sync
                 final Map<StreamIdentifier, List<Lease>> streamToLeasesMap = getStreamToLeasesMap(
-                        currentStreamConfigMap.keySet());
+                        streamConfigMap.keySet());
 
                 // For each of the stream, check if shard sync needs to be done based on the leases state.
-                for (Map.Entry<StreamIdentifier, StreamConfig> streamConfigEntry : currentStreamConfigMap.entrySet()) {
+                for (Map.Entry<StreamIdentifier, StreamConfig> streamConfigEntry : streamConfigMap.entrySet()) {
+                    // Skip shardSync if CSCM has purged the stream.
+                    if (!currentStreamConfigMap.containsKey(streamConfigEntry.getKey())) {
```

Comment:
+1 for short-circuit in scenario where `currentStreamConfigMap` is no longer aware of the stream.

However, to reduce the wall-time between "check" and "execute", WDYT about moving this short-circuit to approximately line ~203 just-before `shardSyncTaskManagerProvider.apply(...)`. This may yield the highest rate of short-circuits which avoid a network call.

Example:
```
log.info("Periodic shard ...");
final StreamConfig streamConfig = currentStreamConfigMap.get(streamConfigEntry.getKey());
if (streamConfig == null) {
   // Skip shardSync if CSCM has purged the stream.
   continue;
}
final ShardSyncTaskManager sSTM = sSTMP.apply(streamConfig);
// rest of block is unchanged
```

Reply:
That's a good callout. I think we can keep the current one and add this new check to optimize both the CPU time and network bandwidth and to achieve higher consistency.